webrtc.html

```diff
@@ -10898,92 +10886,82 @@ <h2 id="sec.idp-error-handling">IdP Error Handling</h2>
-    pc.onnegotiationneeded = function () {
-        pc.createOffer().then(function (offer) {
-            return pc.setLocalDescription(offer);
-        })
-        .then(function () {
-            // send the offer to the other peer
-            signalingChannel.send(JSON.stringify({ desc: pc.localDescription }));
-        })
-        .catch(logError);
-    };
+          const signalingchannel = new signalingchannel();
+          const configuration = { iceservers: [{ urls: "stuns:stun.example.org" }] };
+          let pc;
+          let channel;
 
-    if (isInitiator) {
-        // create data channel and setup chat
-        channel = pc.createDataChannel("chat");
-        setupChat();
-    } else {
-        // setup chat on incoming data channel
-        pc.ondatachannel = function (evt) {
-            channel = evt.channel;
-            setupChat();
-        };
-    }
-}
+          // call start(true) to initiate
+          function start(isinitiator) {
+              pc = new rtcpeerconnection(configuration);
 
-signalingChannel.onmessage = function (evt) {
-    if (!pc)
-        start(false);
+              // send any ice candidates to the other peer
+              pc.onicecandidate = (evt) =&gt; {
+                  signalingchannel.send(json.stringify({ candidate: evt.candidate }));
+              };
 
-    var message = JSON.parse(evt.data);
-    if (message.desc) {
-        var desc = message.desc;
+              // let the "negotiationneeded" event trigger offer generation
+              pc.onnegotiationneeded = async () =&gt; {
+                  try {
+                      const offer = await pc.createoffer();
+                      await pc.setlocaldescription(offer);
+                      // send the offer to the other peer
+                      signalingchannel.send(json.stringify({ desc: pc.localdescription }));
+                  } catch (err) {
+                      console.error(err);
+                  }
+              };
 
-        // if we get an offer, we need to reply with an answer
-        if (desc.type == "offer") {
-            pc.setRemoteDescription(desc).then(function () {
-                return pc.createAnswer();
-            })
-            .then(function (answer) {
-                return pc.setLocalDescription(answer);
-            })
-            .then(function () {
-                var str = JSON.stringify({ desc: pc.localDescription });
-                signalingChannel.send(str);
-            })
-            .catch(logError);
-        } else
-            pc.setRemoteDescription(desc).catch(logError);
-    } else
-        pc.addIceCandidate(message.candidate).catch(logError);
-};
+              if (isinitiator) {
+                  // create data channel and setup chat
+                  channel = pc.createdatachannel("chat");
+                  setupchat();
+              } else {
+                  // setup chat on incoming data channel
+                  pc.ondatachannel = (evt) =&gt; {
```

Comment:
Not in this PR, but this reminds me we really should show an example for `negotiated` data channels as well, maybe even instead of this. cc @taylor-b